File: oop-structs/go/dog.go
package main

import (
	"fmt"
	"slices"
)

const DESCRIPTION string = "This is a package that describes a dog."

var BEST_DOG_BREEDS = []string{"Golden Retriever", "Border Collie"}

type Dog struct {
	name  string
	breed string
}

func newDog(name string, breed string) *Dog {
	return &Dog{name, breed}
}

func (dog *Dog) isBestBreed() bool {
	return slices.Contains(BEST_DOG_BREEDS, dog.breed)
}

func (dog *Dog) isBestDog() {
	if dog.isBestBreed() {
		fmt.Printf("%s is the best dog, congrats!\n", dog.name)
	} else {
		fmt.Printf("%s is not the best dog, sorry.\n", dog.name)
	}
}

func Description() {
	fmt.Println(DESCRIPTION)
}

func main() {
	Description()
	var noah *Dog = newDog("Noah", "Golden Retriever")
	var nina *Dog = newDog("Nina", "Border Collie")
	var pepe *Dog = newDog("Pepe", "Pitbull")
	noah.isBestDog()
	nina.isBestDog()
	pepe.isBestDog()
}

File: oop-structs/ruby/dog.rb
class Dog
  BEST_DOG_BREEDS = ['Golden Retriever', 'Border Collie'].freeze
  def initialize(name, breed)
    @breed = breed
    @name = name
  end

  def self.description
    puts "This is a class that describes a Dog."
  end

  def is_a_good_dog
    if best_breed?
      puts "#{@name} is the best dog ever, congrats!"
    else
      puts "#{@name} is not the best dog, sorry."
    end
  end

  private

  attr_reader :breed, :name
  def best_breed?
    BEST_DOG_BREEDS.include?(@breed)
  end
end

Dog.description
noah = Dog.new('Noah', 'Golden Retriever')
nina = Dog.new('Nina', 'Border Collie')
pepe = Dog.new('Pepe', 'Pitbull')

noah.is_a_good_dog
nina.is_a_good_dog
pepe.is_a_good_dog

